sw/device/lib/drivers/dma/dma.c

```diff
@@ -270,58 +283,130 @@ static struct
+                return;
+            }
+            else if (i > DMA_HP_INTR_INDEX)
+            {
+                dma_hp_tr_intr_counter = 0;
+            }
+
+            #else
+
+            if (i <= DMA_HP_INTR_INDEX)
+            {
+                return;
+            }
+            #endif
+
+            #endif
+        }
+    }
+    return;
 }
 
-void dma_init( dma *peri )
+void dma_init( dma *dma_peri )
 {
     /*
      * If a DMA peripheral was provided, use that one, otherwise use the
      * integrated one.
      */
-    dma_cb.peri = peri ? peri : dma_peri;
-
-    /* Clear the loaded transaction */
-    dma_cb.trans = NULL;
-    /* Clear all values in the DMA registers. */
-    dma_cb.peri->SRC_PTR        = 0;
-    dma_cb.peri->DST_PTR        = 0;
-    dma_cb.peri->ADDR_PTR       = 0;
-    dma_cb.peri->SIZE_D1        = 0;
-    dma_cb.peri->SIZE_D2        = 0;
-    dma_cb.peri->SRC_PTR_INC_D1 = 0;
-    dma_cb.peri->SRC_PTR_INC_D2 = 0;
-    dma_cb.peri->DST_PTR_INC_D1 = 0;
-    dma_cb.peri->DST_PTR_INC_D2 = 0;
-    dma_cb.peri->DIM_CONFIG     = 0;
-    dma_cb.peri->SLOT           = 0;
-    dma_cb.peri->SRC_DATA_TYPE  = 0;
-    dma_cb.peri->DST_DATA_TYPE  = 0;
-    dma_cb.peri->SIGN_EXT       = 0;
-    dma_cb.peri->MODE           = 0;
-    dma_cb.peri->WINDOW_SIZE    = 0;
-    dma_cb.peri->INTERRUPT_EN   = 0;
-    dma_cb.peri->PAD_TOP        = 0;
-    dma_cb.peri->PAD_BOTTOM     = 0;
-    dma_cb.peri->PAD_LEFT       = 0;
-    dma_cb.peri->PAD_RIGHT      = 0;
-    dma_cb.peri->DIM_INV        = 0;
+
+    for (int i = 0; i < DMA_CH_NUM; i++)
+    {
+        dma_subsys_per[i].peri = dma_peri ? dma_peri : dma_peri(i);
+
+        /* Clear the loaded transaction */
+        dma_subsys_per[i].trans = NULL;
+
+        /* Clear all values in the DMA registers. */
+        dma_subsys_per[i].peri->SRC_PTR        = 0;
+        dma_subsys_per[i].peri->DST_PTR        = 0;
+        dma_subsys_per[i].peri->SIZE_D1        = 0;
+        dma_subsys_per[i].peri->SIZE_D2        = 0;
+        dma_subsys_per[i].peri->SRC_PTR_INC_D1 = 0;
+        dma_subsys_per[i].peri->SRC_PTR_INC_D2 = 0;
+        dma_subsys_per[i].peri->DST_PTR_INC_D1 = 0;
+        dma_subsys_per[i].peri->DST_PTR_INC_D2 = 0;
+        dma_subsys_per[i].peri->DIM_CONFIG     = 0;
+        dma_subsys_per[i].peri->DIM_INV        = 0;
+        dma_subsys_per[i].peri->SLOT           = 0;
+        dma_subsys_per[i].peri->SRC_DATA_TYPE  = 0;
+        dma_subsys_per[i].peri->DST_DATA_TYPE  = 0;
+        dma_subsys_per[i].peri->SIGN_EXT       = 0;
+        dma_subsys_per[i].peri->MODE           = 0;
+        dma_subsys_per[i].peri->WINDOW_SIZE    = 0;
+        dma_subsys_per[i].peri->PAD_TOP        = 0;
+        dma_subsys_per[i].peri->PAD_BOTTOM     = 0;
+        dma_subsys_per[i].peri->PAD_LEFT       = 0;
+        dma_subsys_per[i].peri->PAD_RIGHT      = 0;
+        dma_subsys_per[i].peri->DIM_INV        = 0;
+        dma_subsys_per[i].peri->INTERRUPT_EN   = 0;
```

Comment:
do not forget any new added register

Reply:
I checked and added addr_ptr reset, not new but still important